src/mlpack/methods/grad_boosting/grad_boosting_impl.hpp

```diff
@@ -0,0 +1,221 @@
+    numClasses(0),
+    numModels(0)
+{ /*Nothing to do.*/}
+
+// In case weak learners aren't defined
+/**
+ * Constructor.
+ *
+ * @param data Input data
+ * @param labels Corresponding labels
+ * @param numClasses Number of classes
+ * @param numModels Number of weak learners
+ */
+template<typename MatType>
+GradBoosting<MatType>::
+  GradBoosting(const MatType& data,
+               const arma::Row<size_t>& labels,
+               const size_t numClasses,
+               const size_t numModels) :
+  numClasses(numClasses),
+  numModels(numModels)
+{
+  adjustments.resize(numModels);
+  TrainInternal(data, labels, numClasses, numModels);
+}
+
+// In case the user inputs the arguments for the Weak Learner
+/**
+ * Constructor.
+ *
+ * @param data Input data
+ * @param labels Corresponding labels
+ * @param numClasses Number of classes
+ * @param numModels Number of weak learners
+ * @param other Weak Learner, which has been initialized already.
+ */
+template<typename MatType>
+GradBoosting<MatType>::
+GradBoosting(const MatType& data,
+             const arma::Row<size_t>& labels,
+             const size_t numClasses,
+             const size_t numModels,
+             const size_t minimumLeafSize, 
+             const double minimumGainSplit, 
+             const size_t maximumDepth) :
+  numClasses(numClasses),
+  numModels(numModels)
+{
+  adjustments.resize(numModels);
+  TrainInternal(data, labels, numClasses, numModels, 
+                minimumLeafSize, minimumGainSplit, maximumDepth);
+}
+
+template<typename MatType>
+void GradBoosting<MatType>::
+  Train(const MatType& data,
+        const arma::Row<size_t>& labels,
+        const size_t numClasses,
+        const size_t numModels)
+{
+  adjustments.resize(numModels);
+  SetNumModels(numModels);
+  return TrainInternal(data, labels, numClasses, numModels);
+}
+
+template<typename MatType>
+void GradBoosting<MatType>::
+  Train(const MatType& data,
+        const arma::Row<size_t>& labels,
+        const size_t numClasses,
+        const size_t numModels,
+        const size_t minimumLeafSize, 
+        const double minimumGainSplit, 
+        const size_t maximumDepth)
+{
+  adjustments.resize(numModels);
+  SetNumModels(numModels);
+  return TrainInternal(data, labels, numClasses, numModels,
+                       minimumLeafSize, minimumGainSplit, maximumDepth);
+}
+
+// Classify the given test point.
+template<typename MatType>
+template<typename VecType>
+size_t GradBoosting<MatType>::Classify(const VecType& point) 
+{
+  size_t prediction;
+  Classify<arma::colvec>(point, prediction);
+
+  return prediction;
+}
+
+template<typename MatType>
+template<typename VecType>
+void GradBoosting<MatType>::Classify(const VecType& point,
+                                      size_t& prediction)
+{
+  int tempPrediction = 0;
+
+  for (size_t i = 0; i < weakLearners.size(); ++i) 
+  {
+    size_t p;
+
+    arma::vec tempProb(numClasses, arma::fill::zeros); /*Will not use*/
```

Comment:
We should move this outside the for loop, so we don't create a new temporary object for each learner.

Reply:
Implemented this :+1: